### PR TITLE
React 16

### PR DESCRIPTION
**Pull request by JakeLaCombe**
### Summary
React 16 Uplift. For migration strategies, consult https://github.com/cerner/terra-core/wiki/React-16-Migration-Guide. 

Note: terra-clinical-action-header needs to be uplifted to use core dependencies before the full merge can happen. That's the only change needed for this component. It is fully compatible with React 16, but peer dependencies are throwing console warnings.

Deployment link:
https://terra-framework-deployed-pr-73.herokuapp.com/static/#/site/components

### CODE REVIEW

File: packages/terra-layout/themeable-variables.json [hunk 1]
```diff
@@ -0,0 +1,174 @@
+{
```
Comment (emilyrohrbough) [inline packages/terra-layout/themeable-variables.json]: This file needs removed

Comment (bjankord) [reply packages/terra-layout/themeable-variables.json]: Do we know what causes this, I've seen the themable-variables.json generated in a few different places than expected.

Comment (emilyrohrbough) [reply packages/terra-layout/themeable-variables.json]: Yes. I fixed this issue [here](https://github.com/cerner/terra-site/pull/15/files#diff-ce225c4e66755af34002ffacbbe1cc4bR40) in the themeable-variables plugin in terra-site

**Comment by emilyrohrbough** [comment]
@JakeLaCombe. Terra-site requires a peer dependency of react 15. I have this PR out to open the dependencies to allow for 15 or 16. shttps://github.com/cerner/terra-site/pull/23

**Comment by JakeLaCombe** [comment]
Looks like I have a few other dependencies I need to uplift as well to React 16 that just got merged in. I'll get those updated.

**Comment by JakeLaCombe** [comment]
For this branch, I request not to hit the Update branch button. All of the changes into master will have merge conflicts with the wdio that must be fixed before merging, even if it's a simple button press.

File: package.json [hunk 2]
```diff
@@ -97,25 +98,26 @@
...
-    "react-router": "^3.0.5",
-    "react-router-dom": "^4.0.0",
-    "react-test-renderer": "^15.5.4",
+    "nightwatch": "^0.9.19",
+    "react": "^16.2.0",
+    "react-dom": "^16.2.0",
+    "react-router": "^3.2.0",
+    "react-router-dom": "^4.2.2",
+    "react-test-renderer": "^16.2.0",
+    "raf": "^3.4.0",
```
Comment (benbcai) [inline package.json]: raf should be moved up to line 102.

File: package.json [hunk 3]
```diff
@@ -97,25 +98,26 @@
...
     "stylelint-scss": "^2.0.0",
     "stylelint-suitcss": "^1.0.0",
-    "terra-props-table": "^1.0.0",
+    "terra-props-table": "^2.0.0",
     "terra-site": "^2.0.0-RC.3",
     "terra-toolkit": "^2.7.0",
     "wdio-mocha-framework": "^0.5.11",
     "wdio-visual-regression-service": "^0.8.0",
-    "webdriverio": "4.9.9",
+    "webdriverio": "4.9.8",
```
Comment (benbcai) [inline package.json]: What's the reason for bumping down the minor version?

File: packages/terra-application-header-layout/package.json [hunk 4]
```diff
@@ -24,18 +24,18 @@
...
   "peerDependencies": {
-    "react": "^15.4.2",
-    "react-dom": "^15.4.2",
-    "terra-base": "^2.5.0"
+    "react": "^16.2.0",
+    "react-dom": "^16.2.0",
+    "terra-base": "^3.0.0"
   },
   "devDependencies": {
     "terra-props-table": "^1.10.1"
```
Comment (benbcai) [inline packages/terra-application-header-layout/package.json]: terra-props-table should be bumped to 2.0.0.

File: packages/terra-application-links/package.json [hunk 5]
```diff
@@ -26,19 +26,19 @@
     "terra-props-table": "^1.10.1"
```
Comment (benbcai) [inline packages/terra-application-links/package.json]: terra-props-table should be bumped to 2.0.0.

**Comment by emilyrohrbough** [comment]
I think you need to add yourself to the contributors list.!

File: packages/terra-application-name/package.json [hunk 6]
```diff
@@ -22,20 +22,20 @@
...
   "dependencies": {
     "classnames": "^2.2.5",
     "prop-types": "^15.5.8",
     "terra-app-delegate": "^1.11.0",
     "terra-application-header-layout": "^1.2.0",
     "terra-application-menu-layout": "^1.2.0",
-    "terra-base": "^2.0.0",
+    "terra-base": "^3.0.0",
     "terra-image": "^1.15.0",
     "terra-responsive-element": "^1.15.0"
```
Comment (benbcai) [inline packages/terra-application-name/package.json]: Should terra-image and terra-responsive-element be bumped?

File: packages/terra-navigation-layout/package.json [hunk 7]
```diff
@@ -22,30 +22,30 @@
   },
   "homepage": "https://github.com/cerner/terra-framework#readme",
   "devDependencies": {
-    "react": "^15.4.2",
+    "react": "^16.2.0",
     "react-router-dom": "^4.0.0",
```
Comment (benbcai) [inline packages/terra-navigation-layout/package.json]: Should react-router-dom be 4.2.2?

File: packages/terra-application-links/package.json [hunk 8]
```diff
@@ -26,19 +26,19 @@
...
   },
   "peerDependencies": {
-    "react": "^15.4.2",
-    "react-dom": "^15.4.2",
+    "react": "^16.2.0",
+    "react-dom": "^16.2.0",
     "react-router-dom": "^4.0.0",
```
Comment (benbcai) [inline packages/terra-application-links/package.json]: Should react-router-dom be 4.2.2?

File: packages/terra-navigation-layout/package.json (hunk 7 above)
Comment (JakeLaCombe) [reply packages/terra-navigation-layout/package.json]: I'll uplift them both.

File: postcss.config.js [hunk 9]
```diff
@@ -0,0 +1,3 @@
+/* eslint-disable import/no-extraneous-dependencies */
```
Comment (bjankord) [inline postcss.config.js]: Why the blank postcss.config.js file? What is this being used for?

Comment (bjankord) [reply postcss.config.js]: Talked with @JakeLaCombe offline and he mentioned there is an [issue logged](https://github.com/cerner/terra-site/issues/35) for this.

**Comment by JakeLaCombe** [comment]
If you inspect the console, you will see warnings related to focusable props and isSelectable. These are concerns related to terra-site, and terra-framework will be updated when site is uplifted to React 16

**Comment by mjhenkes** [comment]
Terra-site + Terra-Framework introduces a chicken-egg problem. Terra Site can't be upgraded to React-16 until terra-framework is upgraded to react-16 because terra-site consumes terra-framework components. I propose we upgrade and release terra-framework with console warnings then upgrade terra-site.

**Comment by mmalaker** [comment]
Functional verification completed. The framework site examples display and function correctly after consuming the React 16 changes.

**Pull request merged**